flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/retry/strategy/RetryOnExceptionStrategy.java

```diff
@@ -37,6 +40,12 @@ public RetryOnExceptionStrategy(
     @Override
     public void handleException(String testName, int attemptIndex, Throwable throwable)
             throws Throwable {
+        // Failed when reach the total retry times
+        if (attemptIndex >= totalTimes) {
+            LOG.error("Test Failed at the last retry.", throwable);
+            throw throwable;
+        }
```

Comment:
I'm wondering whether we should move this code into `AbstractRetryStrategy` instead of using the same code snippet in both implementing classes.

Reply:
Hello!  I thought about this too -- the way the code is expressed, there could be other implementations of `AbstractRetryStrategy` that apply different strategies (for example, to continue retrying for max period of time regardless of the number of attempts).

My opinion isn't strong on this, but the logic isn't really complicated enough to need to be abstracted away.

Reply:
Valid point - but your example would probably get its own implementation of `RetryStrategy` instead of relying on `AbstractRetryStrategy` as the parent. A counter-argument would be to not try to optimize code for future feature to keep the code as small as possible. If someone decides to come up with a feature that doesn't fit the current implementation, it's time to refactor. Having code duplicates usually doesn't help but introduces potential problems, e.g. bugs only being fixed in one version but not the other.

Anyway, I still see your point...

Reply:
I leave this up to you. No strong opinion here since, as you said, it's a small code change. So, I consider it an edge case. But in general, my comment from above still applies: Duplicate code creates the risk of fixing bugs only in one location but not the other.